src/mkdocs_autorefs/references.py

```diff
@@ -197,9 +203,61 @@ def fix_refs(html: str, url_mapper: Callable[[str], str]) -> tuple[str, list[str
     return html, unmapped
 
 
+class AnchorScannerTreeProcessor(Treeprocessor):
+    """Tree processor to scan and register HTML anchors."""
+
+    _htags: ClassVar[set[str]] = {"h1", "h2", "h3", "h4", "h5", "h6"}
+
+    def __init__(self, plugin: AutorefsPlugin, md: Markdown | None = None) -> None:
+        """Initialize the tree processor.
+
+        Parameters:
+            plugin: A reference to the autorefs plugin, to use its `register_anchor` method.
+        """
+        super().__init__(md)
+        self.plugin = plugin
+
+    def run(self, root: Element) -> None:  # noqa: D102
+        if self.plugin.current_page is not None:
+            self._scan_anchors(root)
+
+    @staticmethod
+    def _slugify(value: str, separator: str = "-") -> str:
+        value = unicodedata.normalize("NFKD", str(value)).encode("ascii", "ignore").decode("ascii")
+        value = re.sub(r"[^\w\s-]", "", value.lower())
+        return re.sub(r"[-_\s]+", separator, value).strip("-_")
+
+    def _scan_anchors(self, parent: Element) -> str | None:
+        hid = None
+        for el, next_el in zip_longest(parent, parent[1:], fillvalue=Element("/")):
+            if el.tag == "a":
+                hid = el.get("id")
+            elif el.tag == "p" and (hid := self._scan_anchors(el)):
+                href = (next_el.get("id") or self._slugify(next_el.text or "")) if next_el.tag in self._htags else ""
+                self.plugin.register_anchor(self.plugin.current_page, hid, href)  # type: ignore[arg-type]
+                hid = None
+            else:
+                hid = self._scan_anchors(el)
+        return hid
+
+
 class AutorefsExtension(Extension):
     """Extension that inserts auto-references in Markdown."""
 
+    def __init__(
+        self,
+        anchor_scanner_factory: Callable[[Markdown], AnchorScannerTreeProcessor] | None = None,
```

Comment:
...changing that to:

```suggestion
        plugin: AutorefsPlugin | None = None,
```